src/main/java/sparksoniq/jsoniq/runtime/iterator/SparkRuntimeIterator.java

```diff
@@ -21,111 +21,54 @@
 package sparksoniq.jsoniq.runtime.iterator;
 
 import org.apache.spark.api.java.JavaRDD;
-import org.apache.spark.sql.Dataset;
-import org.apache.spark.sql.Row;
 import org.rumbledb.api.Item;
-import sparksoniq.Main;
-import sparksoniq.exceptions.IteratorFlowException;
 import sparksoniq.exceptions.SparkRuntimeException;
 import sparksoniq.io.json.JiqsItemParser;
-import sparksoniq.io.json.RowToItemMapper;
 import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
 import sparksoniq.semantics.DynamicContext;
-import sparksoniq.spark.SparkSessionManager;
 
 import java.util.List;
 
-public abstract class SparkRuntimeIterator extends RuntimeIterator {
-
+public abstract class SparkRuntimeIterator extends HybridRuntimeIterator {
 
     private static final long serialVersionUID = 1L;
-    protected JiqsItemParser parser;
-    protected JavaRDD<Item> _rdd;
-    protected List<Item> result = null;
-    protected int currentResultIndex = 0;
 
     protected SparkRuntimeIterator(List<RuntimeIterator> children, IteratorMetadata iteratorMetadata) {
         super(children, iteratorMetadata);
         this.parser = new JiqsItemParser();
     }
 
     @Override
-    public boolean isRDD() {
+    protected boolean initIsRDD() {
         return true;
     }
 
-    @Override
-    public boolean isDataFrame() {
-        return false;
-    }
-
-    @Override
-    public void reset(DynamicContext context) {
-        super.reset(context);
-        result = null;
+    protected JavaRDD<Item> getRDDAux(DynamicContext context) {
```

Comment:
How about moving the exception-throwing implementations of getRDDAux, openLocal, closeLocal, etc to the HybridRuntimeIterator class, so that all iterators (including this one) automatically throw these exceptions by default, without having to implement them if they do not need them?

Reply:
I'm really glad you've asked this because I have considered this as well. Leaving them abstract gives us static errors which I thought might be helpful during development as new iterators are added since developer will exactly know which methods he/she has to implement. Moving exceptions to HybridRI will serve the same purpose as well with less code, but errors for forgotten method implementations will only pop-up at runtime. I don't really have a strong preference. How should we proceed?

Reply:
Both approaches make sense. It is a compromise between on the one hand, assurance not to forget any implementation, and on the other hand, productivity.

My feeling is that, since we have tests for pretty much any iterator, any error that would no longer be thrown statically would otherwise be caught by the tests, which makes me lean towards productivity. It seems we can get the best of both worlds.

Does it make sense?

Reply:
Your reasoning makes sense as well. I like it.

Here is a suggestion along the lines of your latest proposal: keep your implementation the way it is, but rename the SparkRuntimeIterator to RDDRuntimeIterator, and use this iterator as the base class of all iterators that can always return RDDs.

Likewise, you can have an iterator DataFrameRuntimeIterator that always returns true for isDataFrames, and throws an exception for getRDDAux. This iterator could in particular be the base class of ParquetFileblabla, as this one is the typical example of an iterator that can always return a DataFrame.

Would this work?

Reply:
I had the same idea as well but as I'm not very familiar with the performance analysis of method lookups across subclasses I was hesitant to grow the subclass chain further.  So I didn't follow through with this at that moment thinking it might be a premature organization, so to say, in the code as we have only 3-4 SparkOnly (RDD or DF) methods so far.

If a latency is not of a concern here, I really like this idea.

Reply:
Thanks a lot for having performance in mind, this is a very good thought. I agree. Let us keep the DFRuntimeIterator idea aside for now and proceed with only the RDDRuntimeIterator.

(My response to you for later: you do have a point that too big hierarchies and too many function calls are bad for performance. Overall I am especially concerned about the latency of the local iterator methods, because these are the ones called 1000000s of times. The RDD/DF methods are only called once per sequence, and the latency there is less likely to be the bottleneck. In the very-long term, the design with the fastest potential is having a flat hierarchy with RuntimeIterator as an interface with all iterators directly implementing it, with the hybrid code called with a has-a rather than an is-a mechanism. This is something that we can have a discussion on in the long-term.)

Reply:
I like the name RDDRuntimeIterator, it brings much better clarity.

Regarding making RDDAux abstract in RDDRI however, would affect productivity negatively I think. In that scenerio, for every DF only iterator, we would have to manually add the identical exception throwing method for getRDDAux. Eg. ParquetFunctionIterator would need to define this method explicitly.

Reply:
Perfect. Very clear thinking. I'll just mark as approve and you can go ahead.